src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,500 @@
+#include "jni_utils.hpp"
+
+namespace rapids {
+namespace jni {
+
+/**
+ * Convert a string to lower case. It uses std::tolower per character which has limitations
+ * and may not produce the exact same result as the JVM does. This is probably good enough
+ * for now.
+ */
+std::string unicode_to_lower(std::string & input) {
+  int wide_size = std::mbstowcs(nullptr, input.data(), 0);
+  if (wide_size < 0) {
+    throw std::invalid_argument("invalid character sequence...");
+  }
+  std::vector<wchar_t> wide(wide_size + 1);
+  std::mbstowcs(wide.data(), input.data(), wide_size);
+  for (auto wit = wide.begin(); wit != wide.end(); wit++) {
+    *wit = std::towlower(*wit);
+  }
+  int mb_size = std::wcstombs(nullptr, wide.data(), 0);
+  std::string ret(mb_size, '\0');
+  std::wcstombs(ret.data(), wide.data(), mb_size);
+  return ret;
+}
+
+/**
+ * Holds a set of "maps" that are used to rewrite various parts of of the parquet metadata.
+ * Generally each "map" is a gather map that pulls data from an input vector to be placed in
+ * an output vector.
+ */
+struct column_pruning_maps {
+  // gather map for pulling out items from the schema
+  std::vector<int> schema_map;
+  // Each SchemaElement also include the number of children in it. This allows the vector
+  // to be interpreted as a tree flattend depth first. These are the new values for num
+  // children after the schema is gathered.
+  std::vector<int> schema_num_children;
+  // There are several places where a struct is stored only for a leaf column (like a column chunk)
+  // This holds the gather map for those cases.
+  std::vector<int> chunk_map;
+};
+
+/**
+ * This class will handle processing column pruning for a schema. It is written as a class because
+ * of JNI we are sending the names of the columns as a depth first list, like parquet does internally.
+ */
+class column_pruner {
+public:
+    /**
+     * Create pruining filter from a depth first flattened tree of names and num_children.
+     * The root entry is not included in names or in num_children, but parent_num_children
+     * should hold how many entries there are in it.
+     */
+    column_pruner(const std::vector<std::string> & names, 
+            const std::vector<int> & num_children, 
+            int parent_num_children): children(), s_id(0), c_id(-1) {
+      add_depth_first(names, num_children, parent_num_children);
+    }
+
+    column_pruner(int s_id, int c_id): children(), s_id(s_id), c_id(c_id) {
+    }
+
+    column_pruner(): children(), s_id(0), c_id(-1) {
+    }
+
+    /**
+     * Given a schema from a parquet file create a set of pruning maps to prune columns from the rest of the footer
+     */
+    column_pruning_maps filter_schema(std::vector<parquet::format::SchemaElement> & schema, bool ignore_case) {
+      CUDF_FUNC_RANGE();
+      // The maps are sorted so we can compress the tree...
+      // These are the outputs of the computation
+      std::map<int, int, std::less<int>> chunk_map;
```

Comment:
mostly curious, why do you need to specify `std::less<int>` as the comparator? that looks to be the default for `std::map`.

Reply:
It is I just wanted to be explicit that this is a sorted map, and that less gave us the desired order.